feat: add w3c trace id format support

ishajos opened the pull request:
Rev3
- removed setting tracestate header as it is optional
- used the existing segment id instead of creating a new one for this operation to match xray tracing behavior
- added a new smoke test file`smoke_w3c_format_enabled.html` and updated the smoke tests


Rev2
- added integ test
- updated smoke test
- added more unit tests as per suggestions
- updated docs as per suggestions

Add w3c trace id format support to web client
- Add `enableW3CTraceId` option in configuration
- Make changes in FetchPlugin and XhrPlugin to 
     - use w3c trace id format when enabled
     - use `traceparent` header instead of `X-Amzn-Trace-Id` when `addXRayTraceIdHeader` is enabled 

### Modified behavior
- When `enableW3CTraceId` is set to true:
    - check for `traceparent` header and if it is valid
        - update segment id
        - send xray trace event with w3c trace id
    - if no `traceparent` header
        - create a new trace id in w3c format and add to header if `addXRayTraceIdHeader` is enabled 

More information about w3c trace id format : https://www.w3.org/TR/trace-context/

---

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

williazz reviewed (commented):
Issue: I believe you have to update the smoke tests as well. If there are pending backend changes, then please just follow up with that change.

williazz commented on src/plugins/utils/http-utils.ts:
```diff
@@ -189,14 +211,46 @@ export const getAmznTraceIdHeaderValue = (
+    return '00-' + traceId + '-' + segmentId + '-01';
+};
+
+export const isValidW3CHeader = (headerComponents: string[]) => {
```
Question: can we enforce safety for null pointer exceptions here?

williazz commented on src/plugins/utils/http-utils.ts:
```diff
@@ -163,23 +166,42 @@ export const requestInfoToHostname = (request: Request | URL | string) => {
 ) => {
-    headers.set(X_AMZN_TRACE_ID, getAmznTraceIdHeaderValue(traceId, segmentId));
+    if (w3cTraceIdEnabled) {
+        headers.set(
```
Can we cover this line in the integration tests?

williazz commented on src/plugins/event-plugins/XhrPlugin.ts:
```diff
@@ -346,13 +349,23 @@ export class XhrPlugin extends MonkeyPatched<XMLHttpRequest, 'send' | 'open'> {
+                                    xhrDetails.trace!.subsegments![0].id
+                                )
+                            );
+                        }
```
ack nvm then

ishajos commented on src/plugins/event-plugins/__tests__/FetchPlugin.test.ts:
```diff
@@ -1,11 +1,17 @@
 import { FetchPlugin } from '../FetchPlugin';
-import { HttpPluginConfig, X_AMZN_TRACE_ID } from '../../utils/http-utils';
+import {
```
correct me if i'm wrong but i think this must be happening even now
for example if a request contains the traceparent header and rn we only check for the xray header and don't actively remove any other headers

similarly now we only care for the one which is enabled in the config and there might be cases where both headers exist.
i think the backend only cares about the trace id in the xray rum event

limhjgrace replied:
is it expected for a request to have both? but yeah if it doesn't affect backend, shouldn't be an issue.

ishajos commented on src/plugins/event-plugins/__tests__/FetchPlugin.test.ts:
```diff
@@ -515,7 +647,31 @@ describe('FetchPlugin tests', () => {
+        });
+    });
+
+    test('when url matches no regex in addXRayTraceIdHeader then X-Amzn-Trace-Id header is not added to the HTTP request', async () => {
```
yeah good call i'll add that

ishajos commented on src/plugins/utils/http-utils.ts:
```diff
@@ -189,14 +211,46 @@ export const getAmznTraceIdHeaderValue = (
+                ?.split('-');
+            if (headerComponents && isValidW3CHeader(headerComponents)) {
+                traceHeader.traceId = headerComponents[1];
+                traceHeader.segmentId = generateSegmentId();
```
from what i gathered we need to update the segmentid(parent-id) for every outgoing request
https://www.w3.org/TR/trace-context/#a-traceparent-is-received

ishajos replied:
changed it to use the existing segment id to match what we do with xray

limhjgrace commented on src/plugins/utils/http-utils.ts:
```diff
@@ -163,23 +166,42 @@ export const requestInfoToHostname = (request: Request | URL | string) => {
+    if (w3cTraceIdEnabled) {
+        (init.headers as any)[W3C_TRACEPARENT_HEADER_NAME] =
+            getW3CTraceIdHeaderValue(traceId, segmentId);
+        (init.headers as any)[W3C_TRACESTATE_HEADER_NAME] = '';
```
from the docs, it seems like this is optional (doc says may/should) . can we confirm that it is required? i don't see a use case for sending it with an empty value.

ishajos replied:
it needs to be updated when traceparent is updated

anubmand replied:
Is setting empty the usual practice if we dont require any specific tracestate?

ishajos replied:
on reading the docs again i have removed the places where tracestate was being initialized.

limhjgrace commented on smoke/smoke-test-application-NPM-CJS/src/loader-npm-rum-2.ts:
```diff
@@ -14,7 +14,8 @@ try {
         },
-        useBeacon: false
+        useBeacon: false,
+        enableW3CTraceId: true
```
suggestion: we could just enable one of the loader files to have w3c traceId enabled so that we can test x-ray events for header types.

limhjgrace commented on smoke/smoke-test-application-NPM-ES/src/loader-npm-rum-2.ts:
```diff
@@ -15,7 +15,8 @@ try {
         useBeacon: false,
-        releaseId: '2.1.7'
+        releaseId: '2.1.7',
+        enableW3CTraceId: true
```
suggestion: same as above for having just one loader file enabled

limhjgrace commented on smoke/smoke-test-application-CDN/smoke.html:
```diff
@@ -40,7 +40,8 @@
                 useBeacon: false,
-                releaseId: '2.1.7'
+                releaseId: '2.1.7',
+                enableW3CTraceId: true
```
question: do we not want to still test for regression in the x-ray trace header? this might be more work since we'll need separate html files to test but i think this might be worth unless the general move is towards w3c traceIds.

ishajos replied:
added a separate html file

limhjgrace commented on src/loader/loader-w3c-format-enabled.js:
```diff
@@ -0,0 +1,24 @@
+    accessKeyId: 'a',
+    secretAccessKey: 'b',
+    sessionToken: 'c'
+});
```
question: are we adding smoke tests for w3c traceIds? i forget if we do specific validations in the event data in smoke tests but if we do, we should add a test to verify the different traceIds.

ishajos replied:
we don't do specific validation for event fields but i have added a separate smoke test for w3c traceids

ishajos commented on src/plugins/utils/http-utils.ts:
```diff
@@ -189,14 +211,46 @@ export const getAmznTraceIdHeaderValue = (
+                traceHeader.traceId = headerComponents[1];
+                traceHeader.segmentId = generateSegmentId();
+            } else {
+                headers.set(W3C_TRACESTATE_HEADER_NAME, '');
```
discussed above

ishajos commented on src/plugins/utils/http-utils.ts:
```diff
@@ -189,14 +211,46 @@ export const getAmznTraceIdHeaderValue = (
+            if (headerComponents && isValidW3CHeader(headerComponents)) {
+                traceHeader.traceId = headerComponents[1];
+                traceHeader.segmentId = generateSegmentId();
+            } else {
```
this is created in here : https://github.com/aws-observability/aws-rum-web/pull/655/files#diff-2c005fdd6d58bbed6617d3d80a8dcc716198780fd01e46fe0f52461085be4a8eL103

in the `addXRayTraceIdHeader` function

ps863 commented on app/smoke_w3c_format_enabled.html:
```diff
@@ -0,0 +1,274 @@
+                cwr('enable');
+            }
+
+            // //////////////////////////
```
nit: remove this

ishajos replied:
done

ps863 commented on docs/configuration.md:
```diff
@@ -134,7 +135,7 @@ telemetries: [
 | stackTraceLength | Number | `1000 ` | The number of characters to record from a JavaScript error's stack trace (if available). |
 | recordAllRequests | boolean | `false` | By default, only HTTP failed requests (i.e., those with network errors or status codes which are not 2xx) are recorded. When this field is `true`, the http telemetry will record all requests, including those with successful 2xx status codes. <br/><br/>This field does **does not apply** to X-Ray traces, where all requests are recorded. |
-| addXRayTraceIdHeader | boolean or RegExp[] | `false` | By default, the `X-Amzn-Trace-Id` header will not be added to the HTTP request. This means that the client-side trace and server-side trace will **not be linked** in X-Ray or the ServiceLens graph.<br/><br/> When this field is `true`, the `X-Amzn-Trace-Id` header will be added to HTTP requests (`XMLHttpRequest` or `fetch`).<br/><br/> When t ...
+| addXRayTraceIdHeader | boolean or RegExp[] | `false` | By default, the `X-Amzn-Trace-Id` header will not be added to the HTTP request. This means that the client-side trace and server-side trace will **not be linked** in X-Ray or the ServiceLens graph.<br/><br/> When this field is `true`, the `X-Amzn-Trace-Id` header will be added to HTTP requests (`XMLHttpRequest` or `fetch`).<br/><br/> When t ...
```
nit: Should trace parent hyperlink to this - https://www.w3.org/TR/trace-context/#traceparent-header?

ishajos replied:
done

ps863 commented on src/plugins/utils/http-utils.ts:
```diff
@@ -189,14 +208,44 @@ export const getAmznTraceIdHeaderValue = (
+};
+
+export const isValidW3CHeader = (headerComponents: string[]) => {
+    return (
```
question: Can you document in the code where these specifications are defined in w3c docs?

ishajos replied:
done

ps863 commented on src/plugins/event-plugins/__tests__/XhrPlugin.test.ts:
```diff
@@ -120,6 +121,60 @@ describe('XhrPlugin tests', () => {
         });
     });
 
+    test('when XHR is called and w3c trace format enabled then the plugin records a trace', async () => {
```
comment: can we or is there already a test that tests the behavior if w3TraceEnabled is true but enableXRay is false?

ishajos replied:
good catch! added tests for this case

ps863 commented on src/plugins/utils/http-utils.ts:
```diff
@@ -189,14 +208,44 @@ export const getAmznTraceIdHeaderValue = (
+export const isValidW3CHeader = (headerComponents: string[]) => {
+    return (
+        headerComponents?.length === 4 &&
+        headerComponents[0].length === 2 &&
```
nit: consider defining these as constants

ishajos replied:
done

ishajos commented:
## Update:

Fixed smoke tests w3c format trace id

- added a new smoke test file smoke_w3c_format_enabled.html
- refactored the getURL function to load different test pages
- updated the steps to update and upload smoke test files so that the new loader file is updated

NOTE : 2 of the smoke tests are remove because they will fail without the backed changes. Will add them back in a later PR and rerun the smoke tests once the changes are deployed

Smoke test results :

williazz commented on src/orchestration/Orchestration.ts:
```diff
@@ -95,6 +95,7 @@ export const defaultConfig = (cookieAttributes: CookieAttributes): Config => {
         telemetries: [],
         useBeacon: true,
         userIdRetentionDays: 30,
+        enableW3CTraceId: false,
```
Should this be enabled in the future? Maybe next major version

ishajos replied:
no this is for the default state. we don't want it enabled by default

williazz reviewed (commented):
Question: Can we clarify what happens in the following 4 scenarios, and add documentation? 

1. W3C header exists + enableW3CTraceId=true => do not begin new trace, override http event with existing w3c header
2. XRay header exists + enableW3CTraceId=false => do not begin new trace, override http event with existing xray header
3. W3C header does not exist + enableW3CTraceId=true => begin new trace, override http event with new w3c header
4. XRay header does not exist + enableW3cTraceId=false => begin new trace, override http event with new xray header


And then this also happens: 

1. If W3C trace is enabled, then any existing XRay trace is ignored. 
2. If W3C trace is disabled, then any existing W3C trace is ignored.
